components/playlist/playlist_tab_helper.cc

```diff
@@ -0,0 +1,61 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/playlist/playlist_tab_helper.h"
+
+namespace playlist {
+
+PlaylistTabHelper::PlaylistTabHelper(content::WebContents* contents)
+    : content::WebContentsObserver(contents),
+      content::WebContentsUserData<PlaylistTabHelper>(*contents) {}
+
+PlaylistTabHelper::~PlaylistTabHelper() = default;
+
+// static
+void PlaylistTabHelper::HideMediaSrcAPIFromContents(
+    content::WebContents* contents) {
+  if (auto url = contents->GetVisibleURL();
+      url.scheme() == "chrome-extension" || url.host_piece() == "127.0.0.1") {
+    // Don't hide API for our own pages. e.g. pages in browser tests
+    return;
+  }
+
+  DCHECK(contents && contents->GetMainFrame());
+
+  // This script is from
+  // https://github.com/brave/brave-ios/blob/development/Client/Frontend/UserContent/UserScripts/PlaylistSwizzler.js
+  static const std::u16string kScriptToHideMediaSourceAPI =
+      uR"-(
+    (function() {
+      // Stub out the MediaSource API so video players do not attempt to use `blob` for streaming
+      if (window.MediaSource || window.WebKitMediaSource || window.HTMLMediaElement && HTMLMediaElement.prototype.webkitSourceAddId) {
+        window.MediaSource = null;
+        window.WebKitMediaSource = null;
+        delete window.MediaSource;
+        delete window.WebKitMediaSource;
+      }
+    })();
+    )-";
+
+  // In order to hide js API from main world, use testing
+  // api temporarily.
+  const auto execute_script_callback = [](std::u16string script,
+                                          content::RenderFrameHost* frame) {
+    frame->ExecuteJavaScriptForTests(kScriptToHideMediaSourceAPI,
```

Comment:
I guess this is fine because it's behind a flag for now (and we were doing this before anyway)

Reply:
Yeah, This is another thing I'm not sure about. AFAIK, ExecutingJavaScript(), which runs in main world, is allowed on Android, but not Desktop. I guess why they allow it for Android is that it isn't supported to make extension API for browser on Android. And iOS folks seems to be doing similar things with WKWebView. So my question is, should we avoid this since/if we can do this with an extension? I'm not sure. @bsclifton @simonhong @petemill  Do you think there's a sort of security reason for preventing ExecuteJavascript in main world on desktop?

Reply:
Yeah, that makes sense :smile: I guess we'll get someone with a bit more authority than me to make that call :+1:

Reply:
Yep, I'm going to ask other folks about this! Also, I'm thinking of making something like `ScopedAllowingJavascriptExecution` which allows JS execution for short time.

Reply:
we do execute js in the main world for instance in wallet, but there are some security concerns like the main page being able to do prototype pollution on the script. cc @bridiver in case he knows a better way here

Reply:
Thanks @diracdeltas . Then can I merge this PR or should I wait for overall review?

Reply:
let me ping brian to take a look first

Reply:
this definitely needs to be changed and the same with the call in PlaylistDownloadRequestManager

Reply:
also it is not going to work reliably anyway. You are making an async call to run JS so you have a race condition. @antonok-edm can we make this work through cf somehow?

Reply:
if we can't enable it through cf, this needs to be a RenderFrameObserver, not a WebContentsObserver

Reply:
May I ask you what `cf` is?

Reply:
It seems we can turn off MediaSourceAPI with RuntimeEnabledFeature. Then we can remove this kind of unreliable js call. Let me try again.

Reply:
I assume `cf` is `cosmetic filters`.
we shouldn't have `*ForTests` anywhere in production code. One of other options is to implement a renderer-based class and call it via mojo to run all playlist-related main frame JS there.
I assume the current approach might break websites in some way, because it's not synchronous, i.e. the time between completed navigation and a requested JS execution might be pretty big.

Reply:
Fortunately, RuntimeEnabledFeature works as we needed. I've removed JS injection part and MediaSource api will be disabled from blink side. But disabling should be improved based on preference or content setting in the future.